src/Elastic.Apm/Agent.cs

```diff
@@ -98,6 +98,39 @@ public static class Agent
 		private static AgentComponents _components;
 		private static volatile bool _isConfigured;
 
+		/// <summary>
+		/// Adds an <see cref="ITransactionObserver"/> instance to the agent, which gets called each time the currently active transaction changes.
+		/// You can only add <see cref="ITransactionObserver"/>s when the agent is already configured - which you can check
+		/// with the <see cref="IsConfigured"/> property.
+		/// </summary>
+		/// <param name="transactionObserver">The instance that will be informed when the active transaction changes</param>
+		/// <returns><code>true</code> if the <param name="transactionObserver"></param> was successfully added, <code>false</code> otherwise.</returns>
+		internal static bool RegisterTransactionObserver(ITransactionObserver transactionObserver)
```

Comment:
Raised this offline as well, it'd be good to have some discussions again about:

https://github.com/elastic/apm-agent-dotnet/pull/537#issuecomment-540230744 and https://github.com/elastic/apm-agent-dotnet/issues/441 in general.

Reply:
My take on it is that `CurrentExecutionSegmentsContainer` is the one that is responsible for this thing - we can move the registration of the observers wherever we want.

The main reason I went with the static `Agent` instance is that I thought that’s a more natural place to register things, plus we can make sure we don’t accidentally new up the `Tracer` - we have an `_isConfigured` check here, plus we can return a `bool`, which is hopefully a good hint for this. 

If we move it to the `Tracer` then users need to check with `Agent.IsConfigured`. I assume users do the registration early in the startup, so we could again have the problem that we new up a tracer without the proper AgentComponents.

But yeah, singleton is bad. Overall I’m ok to move this to `Tracer` and document that people should make sure the agent is configured before they use it.

Reply:
Could you please clarify what do you mean by "before they use it" - aren't all the new methods `internal`?

Reply:
> If we move it to the `Tracer` then users need to check with `Agent.IsConfigured`

If we don't want users to check `Agent.IsConfigured` then why did we make it `public`?

Reply:
Right, in that case in this iteration we need to check it. 

The way I imagined this will be used is that we make the agent `InternalVisibleTo` the corresponding log library integration and in that library we offer a simple method to register it. Somehow we need to make sure users can register it and if there is a chance that this registration fails we should let users know and offer a way to retry.

Reply:
> > If we move it to the `Tracer` then users need to check with `Agent.IsConfigured`
> 
> If we don't want users to check `Agent.IsConfigured` then why did we make it `public`?

If we move it to `Tracer`, then the way you register would be:

```
Agent.Tracer.RegisterTransactionObserver()
```

But that'll new up a tracer, without saying anything, while you would need to check with:

```
Agent.IsConfigured();
```

It's 1 level higher. And that's what I feel is strange. So, yes, we made it public to check if the Agent is configured. My concern is that it's not obvious that you also need to check for `Agent.Tracer.RegisterTransactionObserver` - if it's only `Agent.RegisterTransactionObserve` it feels more natural to me to check on `Agent.IsConfigured`.

But, again, I'm open to do it, I just say try to see what downsides it'd have.

Reply:
One possible approach would be to have users call:
```
if (Agent.IsConfigured) NLogCorrelation.Install();
```
I think it clearer than having to read the docs to find out when `Agent.RegisterTransactionObserver` returns false.

Reply:
> One possible approach would be to have users call:
> 
> ```
> if (Agent.IsConfigured) NLogCorrelation.Install();
> ```
> 
> I think it clearer than having to read the docs to find out when `Agent.RegisterTransactionObserver` returns false.

Yes, I like that - that is almost what I wanted to do. Agree that `RegisterTransactionObserver` should not be public at this point.

So, suggestion:
We move `RegisterTransactionObserver` to `Tracer`, we keep it internal and `NLogCorrelation.Install()` (or whatever we call it) will just set `Agent.Tracer.RegisterTransactionObserver`. Furthermore we make sure `NLogCorrelation.Install()` does not new up a tracer in case there isn't one. 

How does this sound?

Reply:
Of course, all we say in this thread is by assuming we need this PR at all. On that one we'll see the output in https://github.com/NLog/NLog/issues/3726.

Reply:
It sounds good except that I am not sure about

> NLogCorrelation.Install() does not new up a tracer in case there isn't one

It's somewhat of inconsistent with `Agent.Tracer` which forces the agent's singleton to be created but I don't think it's that important to keep `Agent.Tracer` and `NLogCorrelation.Install()` aligned in that regard. Also it means that `NLogCorrelation.Install()` will have to either throw if `Agent.IsConfigured` is `false` or at least return `false`. If we want users to be sure that `Agent.IsConfigured` is `true` before calling `NLogCorrelation.Install()` then throwing seems the better option.

Reply:
Also it's worth sketching out how we expect users to use the log correlation feature with legacy ASP.NET applications where so far we didn't require any changes to application's source code. This aspect seems to be quite important for the customers with legacy .NET applications.